gpkit/constraints/set.py

```diff
@@ -0,0 +1,227 @@
+        subs = substitutions if substitutions else {}
+        if hasattr(self, "cost"):
+            subs.update(self.cost.values)
+        if not isinstance(constraints, ConstraintSet):
+            # constraintsetify everything
+            for i, constraint in enumerate(self):
+                if (hasattr(constraint, "__iter__") and
+                        not isinstance(constraint, ConstraintSet)):
+                    self[i] = ConstraintSet(constraint)
+        else:
+            # grab the substitutions dict from the top constraintset
+            subs.update(constraints.substitutions)
+        self.substitutions = KeyDict.with_keys(self.varkeys,
+                                               self._iter_subs(subs))
+
+    __str__ = _str
+    __repr__ = _repr
+    _repr_latex_ = _repr_latex_
+
+    def str_without(self, excluded=None):
+        "String representation of a ConstraintSet."
+        if not excluded:
+            excluded = ["units"]
+        if "cost" in excluded:
+            lines = []  # start with nothing
+        else:
+            excluded.append("cost")
+            if hasattr(self, "cost"):
+                lines = ["  # minimize",
+                         "        %s" % self.cost.str_without(excluded),
+                         "  # subject to"]
+            else:
+                lines = [""]  # start with a newline
+        for constraint in self:
+            if hasattr(constraint, "_subconstr_str"):
+                cstr = constraint._subconstr_str(excluded)
+                if cstr is None:
+                    cstr = try_str_without(constraint, excluded)
+            else:
+                cstr = try_str_without(constraint, excluded)
+            if cstr[:8] != "        ":  # require indentation
+                cstr = "        " + cstr
+            lines.append(cstr)
+        return "\n".join(lines)
+
+    def latex(self, excluded=None):
+        "LaTeX representation of a ConstraintSet."
+        if not excluded:
+            excluded = ["units"]
+        if "cost" in excluded:
+            lines = []  # start with nothing
+        else:
+            excluded.append("cost")
+            lines = ["\\begin{array}[ll]", "\\text{}"]
+            if hasattr(self, "cost"):
+                lines.extend(["\\text{minimize}",
+                              "    & %s \\\\" % self.cost.latex(excluded),
+                              "\\text{subject to}"])
+        for constraint in self:
+            if hasattr(constraint, "_subconstr_latex"):
+                cstr = constraint._subconstr_latex(excluded)
+                if cstr is None:
+                    cstr = constraint.latex(excluded)
+            else:
+                cstr = constraint.latex(excluded)
+            if cstr[:6] != "    & ":  # require indentation
+                cstr = "    & " + cstr
+            lines.append(cstr)
+        lines.append("\\end{array}")
+        return "\n".join(lines)
+
+    def _subconstr_str(self, excluded=None):
```

Comment:
The idea right now is that it's implemented by subclasses (e.g. ConstraintBase) when they want to simplify their representations. I could see `ArrayConstraints`, gpfits, and taylor expansions implementing it too.